From patchwork Thu Apr  9 13:32:46 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Cathy Avery <cavery@redhat.com>
X-Patchwork-Id: 11481795
Return-Path: <SRS0=jZn3=5Z=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 1A24D913
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 13:33:01 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id EE4BB20857
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 13:33:00 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=redhat.com header.i=@redhat.com
 header.b="aXmWZf5S"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726872AbgDINc7 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 9 Apr 2020 09:32:59 -0400
Received: from us-smtp-delivery-1.mimecast.com ([207.211.31.120]:36514 "EHLO
        us-smtp-1.mimecast.com" rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org
        with ESMTP id S1726641AbgDINc7 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 9 Apr 2020 09:32:59 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=redhat.com;
        s=mimecast20190719; t=1586439178;
        h=from:from:reply-to:subject:subject:date:date:message-id:message-id:
         to:to:cc:mime-version:mime-version:
         content-transfer-encoding:content-transfer-encoding:
         in-reply-to:in-reply-to:references:references;
        bh=woDbbnutdla08US4iOa0Wjr9OpIuMoISMvO6V9wkEkM=;
        b=aXmWZf5Ss2u8RAwLIQLDaUuc+mzQlXIbzlpmIU3w26vIu5Gzlm3/E7Z9oAzCDnV9Me9x1e
        Hh3NSloOMKGD3J/mL+N336uozt3Tzygtx/PD3zRZLUxSg7nzk8C3s7Kn571+mnrw9Ta/CK
        sE7ksVYoCP8yewER+wZZFLQRLwuqjbQ=
Received: from mimecast-mx01.redhat.com (mimecast-mx01.redhat.com
 [209.132.183.4]) (Using TLS) by relay.mimecast.com with ESMTP id
 us-mta-479-ZWniwjOnMzK2rhNlAHaVIQ-1; Thu, 09 Apr 2020 09:32:50 -0400
X-MC-Unique: ZWniwjOnMzK2rhNlAHaVIQ-1
Received: from smtp.corp.redhat.com (int-mx07.intmail.prod.int.phx2.redhat.com
 [10.5.11.22])
        (using TLSv1.2 with cipher AECDH-AES256-SHA (256/256 bits))
        (No client certificate requested)
        by mimecast-mx01.redhat.com (Postfix) with ESMTPS id 8EBC618FF661
        for <kvm@vger.kernel.org>; Thu,  9 Apr 2020 13:32:49 +0000 (UTC)
Received: from localhost.localdomain.com (ovpn-113-69.rdu2.redhat.com
 [10.10.113.69])
        by smtp.corp.redhat.com (Postfix) with ESMTP id 18F4F10027AC;
        Thu,  9 Apr 2020 13:32:49 +0000 (UTC)
From: Cathy Avery <cavery@redhat.com>
To: kvm@vger.kernel.org, pbonzini@redhat.com
Subject: [PATCH kvm-unit-tests 1/2] svm: Add test cases around NMI injection
Date: Thu,  9 Apr 2020 09:32:46 -0400
Message-Id: <20200409133247.16653-2-cavery@redhat.com>
In-Reply-To: <20200409133247.16653-1-cavery@redhat.com>
References: <20200409133247.16653-1-cavery@redhat.com>
MIME-Version: 1.0
X-Scanned-By: MIMEDefang 2.84 on 10.5.11.22
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

This test checks for NMI delivery to L2 and
intercepted NMI (VMEXIT_NMI) delivery to L1.

Signed-off-by: Cathy Avery <cavery@redhat.com>
---
 x86/svm_tests.c | 84 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 84 insertions(+)

diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index 16b9dfd..d1dbdef 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -1340,6 +1340,87 @@ static bool interrupt_check(struct svm_test *test)
     return get_test_stage(test) == 5;
 }
 
+static volatile bool nmi_fired;
+
+#define NMI_VECTOR    2
+
+static void nmi_handler(isr_regs_t *regs)
+{
+    nmi_fired = true;
+    apic_write(APIC_EOI, 0);
+}
+
+static void nmi_prepare(struct svm_test *test)
+{
+    default_prepare(test);
+    nmi_fired = false;
+    handle_irq(NMI_VECTOR, nmi_handler);
+    set_test_stage(test, 0);
+}
+
+static void nmi_test(struct svm_test *test)
+{
+    apic_icr_write(APIC_DEST_SELF | APIC_DEST_PHYSICAL | APIC_DM_NMI | APIC_INT_ASSERT, 0);
+
+    report(nmi_fired, "direct NMI while running guest");
+
+    if (!nmi_fired)
+        set_test_stage(test, -1);
+
+    vmmcall();
+
+    nmi_fired = false;
+
+    apic_icr_write(APIC_DEST_SELF | APIC_DEST_PHYSICAL | APIC_DM_NMI | APIC_INT_ASSERT, 0);
+
+    if (!nmi_fired) {
+        report(nmi_fired, "intercepted pending NMI not dispatched");
+        set_test_stage(test, -1);
+    }
+
+}
+
+static bool nmi_finished(struct svm_test *test)
+{
+    switch (get_test_stage(test)) {
+    case 0:
+        if (vmcb->control.exit_code != SVM_EXIT_VMMCALL) {
+            report(false, "VMEXIT not due to vmmcall. Exit reason 0x%x",
+                   vmcb->control.exit_code);
+            return true;
+        }
+        vmcb->save.rip += 3;
+
+        vmcb->control.intercept |= (1ULL << INTERCEPT_NMI);
+        break;
+
+    case 1:
+        if (vmcb->control.exit_code != SVM_EXIT_NMI) {
+            report(false, "VMEXIT not due to NMI intercept. Exit reason 0x%x",
+                   vmcb->control.exit_code);
+            return true;
+        }
+
+        report(true, "NMI intercept while running guest");
+        break;
+
+    case 2:
+        break;
+
+    default:
+        return true;
+    }
+
+    inc_test_stage(test);
+
+    return get_test_stage(test) == 3;
+}
+
+static bool nmi_check(struct svm_test *test)
+{
+    return get_test_stage(test) == 3;
+}
+
 #define TEST(name) { #name, .v2 = name }
 
 /*
@@ -1446,6 +1527,9 @@ struct svm_test svm_tests[] = {
     { "interrupt", default_supported, interrupt_prepare,
       default_prepare_gif_clear, interrupt_test,
       interrupt_finished, interrupt_check },
+    { "nmi", default_supported, nmi_prepare,
+      default_prepare_gif_clear, nmi_test,
+      nmi_finished, nmi_check },
     TEST(svm_guest_state_test),
     { NULL, NULL, NULL, NULL, NULL, NULL, NULL }
 };

From patchwork Thu Apr  9 13:32:47 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Cathy Avery <cavery@redhat.com>
X-Patchwork-Id: 11481797
Return-Path: <SRS0=jZn3=5Z=vger.kernel.org=kvm-owner@kernel.org>
Received: from mail.kernel.org (pdx-korg-mail-1.web.codeaurora.org
 [172.30.200.123])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 8E88617D4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 13:33:01 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.kernel.org (Postfix) with ESMTP id 6E29F208FE
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu,  9 Apr 2020 13:33:01 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (1024-bit key) header.d=redhat.com header.i=@redhat.com
 header.b="bJVdR+yB"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726873AbgDINdA (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 9 Apr 2020 09:33:00 -0400
Received: from us-smtp-2.mimecast.com ([207.211.31.81]:27765 "EHLO
        us-smtp-delivery-1.mimecast.com" rhost-flags-OK-OK-OK-FAIL)
        by vger.kernel.org with ESMTP id S1726679AbgDINc7 (ORCPT
        <rfc822;kvm@vger.kernel.org>); Thu, 9 Apr 2020 09:32:59 -0400
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=redhat.com;
        s=mimecast20190719; t=1586439178;
        h=from:from:reply-to:subject:subject:date:date:message-id:message-id:
         to:to:cc:mime-version:mime-version:
         content-transfer-encoding:content-transfer-encoding:
         in-reply-to:in-reply-to:references:references;
        bh=TOc6GFFTPKIpzxVPcx4sO1avNIBcz9F7mkm1phzQaq4=;
        b=bJVdR+yBuOr+0865ceJVgz/wsEofM2DNwuELVQMqHxe6yyJpio1lT9vLyh9+9kMRgG0YSl
        3vNEJYg8h+QXDqzTR0BCx/cJSQ1sWN+cdgqKYTYKwBuQkVm5JnraAjMU7sXhD4uCFXHrnR
        Dm3AemFUIVnwHkhAhPo88k20osAHHEA=
Received: from mimecast-mx01.redhat.com (mimecast-mx01.redhat.com
 [209.132.183.4]) (Using TLS) by relay.mimecast.com with ESMTP id
 us-mta-493-G2lCDIYlNomArhFBra1VhA-1; Thu, 09 Apr 2020 09:32:51 -0400
X-MC-Unique: G2lCDIYlNomArhFBra1VhA-1
Received: from smtp.corp.redhat.com (int-mx07.intmail.prod.int.phx2.redhat.com
 [10.5.11.22])
        (using TLSv1.2 with cipher AECDH-AES256-SHA (256/256 bits))
        (No client certificate requested)
        by mimecast-mx01.redhat.com (Postfix) with ESMTPS id 346AA107ACCC
        for <kvm@vger.kernel.org>; Thu,  9 Apr 2020 13:32:50 +0000 (UTC)
Received: from localhost.localdomain.com (ovpn-113-69.rdu2.redhat.com
 [10.10.113.69])
        by smtp.corp.redhat.com (Postfix) with ESMTP id B332710027AC;
        Thu,  9 Apr 2020 13:32:49 +0000 (UTC)
From: Cathy Avery <cavery@redhat.com>
To: kvm@vger.kernel.org, pbonzini@redhat.com
Subject: [PATCH kvm-unit-tests 2/2] svm: Add test cases around NMI injection
 with HLT
Date: Thu,  9 Apr 2020 09:32:47 -0400
Message-Id: <20200409133247.16653-3-cavery@redhat.com>
In-Reply-To: <20200409133247.16653-1-cavery@redhat.com>
References: <20200409133247.16653-1-cavery@redhat.com>
MIME-Version: 1.0
X-Scanned-By: MIMEDefang 2.84 on 10.5.11.22
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

This test checks for NMI delivery to L2 and
intercepted NMI (VMEXIT_NMI) delivery to L1
during an active HLT.

Signed-off-by: Cathy Avery <cavery@redhat.com>
---
 x86/svm_tests.c | 103 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 103 insertions(+)

diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index d1dbdef..7d7c546 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -9,6 +9,7 @@
 #include "alloc_page.h"
 #include "isr.h"
 #include "apic.h"
+#include "delay.h"
 
 #define SVM_EXIT_MAX_DR_INTERCEPT 0x3f
 
@@ -1421,6 +1422,105 @@ static bool nmi_check(struct svm_test *test)
     return get_test_stage(test) == 3;
 }
 
+#define NMI_DELAY 100000000ULL
+
+static void nmi_message_thread(void *_test)
+{
+    struct svm_test *test = _test;
+
+    while (get_test_stage(test) != 1)
+        pause();
+
+    delay(NMI_DELAY);
+
+    apic_icr_write(APIC_DEST_PHYSICAL | APIC_DM_NMI | APIC_INT_ASSERT, id_map[0]);
+
+    while (get_test_stage(test) != 2)
+        pause();
+
+    delay(NMI_DELAY);
+
+    apic_icr_write(APIC_DEST_PHYSICAL | APIC_DM_NMI | APIC_INT_ASSERT, id_map[0]);
+}
+
+static void nmi_hlt_test(struct svm_test *test)
+{
+    long long start;
+
+    on_cpu_async(1, nmi_message_thread, test);
+
+    start = rdtsc();
+
+    set_test_stage(test, 1);
+
+    asm volatile ("hlt");
+
+    report((rdtsc() - start > NMI_DELAY) && nmi_fired,
+          "direct NMI + hlt");
+
+    if (!nmi_fired)
+        set_test_stage(test, -1);
+
+    nmi_fired = false;
+
+    vmmcall();
+
+    start = rdtsc();
+
+    set_test_stage(test, 2);
+
+    asm volatile ("hlt");
+
+    report((rdtsc() - start > NMI_DELAY) && nmi_fired,
+           "intercepted NMI + hlt");
+
+    if (!nmi_fired) {
+        report(nmi_fired, "intercepted pending NMI not dispatched");
+        set_test_stage(test, -1);
+    }
+
+    set_test_stage(test, 3);
+}
+
+static bool nmi_hlt_finished(struct svm_test *test)
+{
+    switch (get_test_stage(test)) {
+    case 1:
+        if (vmcb->control.exit_code != SVM_EXIT_VMMCALL) {
+            report(false, "VMEXIT not due to vmmcall. Exit reason 0x%x",
+                   vmcb->control.exit_code);
+            return true;
+        }
+        vmcb->save.rip += 3;
+
+        vmcb->control.intercept |= (1ULL << INTERCEPT_NMI);
+        break;
+
+    case 2:
+        if (vmcb->control.exit_code != SVM_EXIT_NMI) {
+            report(false, "VMEXIT not due to NMI intercept. Exit reason 0x%x",
+                   vmcb->control.exit_code);
+            return true;
+        }
+
+        report(true, "NMI intercept while running guest");
+        break;
+
+    case 3:
+        break;
+
+    default:
+        return true;
+    }
+
+    return get_test_stage(test) == 3;
+}
+
+static bool nmi_hlt_check(struct svm_test *test)
+{
+    return get_test_stage(test) == 3;
+}
+
 #define TEST(name) { #name, .v2 = name }
 
 /*
@@ -1530,6 +1630,9 @@ struct svm_test svm_tests[] = {
     { "nmi", default_supported, nmi_prepare,
       default_prepare_gif_clear, nmi_test,
       nmi_finished, nmi_check },
+    { "nmi_hlt", smp_supported, nmi_prepare,
+      default_prepare_gif_clear, nmi_hlt_test,
+      nmi_hlt_finished, nmi_hlt_check },
     TEST(svm_guest_state_test),
     { NULL, NULL, NULL, NULL, NULL, NULL, NULL }
 };
